chain/chain/src/chain.rs

```diff
@@ -2088,13 +2088,20 @@ impl Chain {
                     if new_flat_head == CryptoHash::default() {
                         new_flat_head = *self.genesis.hash();
                     }
-                    flat_storage_state.update_flat_head(&new_flat_head).unwrap_or_else(|_| {
-                        panic!(
-                            "Cannot update flat head from {:?} to {:?}",
-                            flat_storage_state.get_flat_head(),
-                            new_flat_head
-                        )
-                    });
+                    let old_flat_head = flat_storage_state.get_flat_head();
+                    // If we have two forks and already postprocessed block from other fork, it can result in new
+                    // proposed flat head behind current flat head. In such case, we don't update flat head.
+                    // Note that we can't take final head because the shard may be catching up.
+                    if self.get_block_header(&old_flat_head)?.height()
```

Comment:
Here we actually checked that the shards are not for the catch up shards through line 2078, so the comment is not exactly right. However, it's still better to use the current flat_head instead of final head because they may not always be the same for other reasons.